File: src/api/tc4Failure.ts
/**
 * tc4Failure.ts — V3.8 TC4 多模态失效分析 API
 *
 * 三路输入：EBSD 晶格 + SEM 图像 + Load 载荷
 * 输出：失效模式分类 + 疲劳寿命预测
 */

import { defineStore } from 'pinia'
import { invoke } from '@tauri-apps/api/core'

// ============ 类型定义 ============

export interface EbsdFeatures {
  sample_id: string
  features: number[]
  a?: number
  c?: number
  c_to_a?: number
}

export interface SemFeatures {
  sample_id: string
  features: number[]
  quality_score: number
}

export interface LoadFeatures {
  sample_id: string
  features: number[]
  load_type: string
}

export type FailureMode = 'HCF' | 'LCF' | 'TMF' | 'CREEP' | 'OVERLOAD' | 'FOD'

export interface MultimodalPrediction {
  sample_id: string
  failure_mode: FailureMode
  class_confidence: number
  log_cycles: number
  cycles: number
  cycle_confidence: number
  fused_features: number[]
}

export interface Tc4SyntheticConfig {
  num_samples: number
  output_dir: string
  use_tc4: boolean
  seed?: number
}

export interface Tc4MdValidationConfig {
  lattice_a: number
  lattice_c: number
  crack_plane: string
  crack_direction: string
  validation_type: string
}

// ============ API 调用 ============

export async function tc4GenerateSynthetic(config: Tc4SyntheticConfig): Promise<string> {
  return invoke('tc4_generate_synthetic', { config })
}

export async function tc4LoadEbsd(ebsdPath: string, material: string): Promise<EbsdFeatures> {
  return invoke('tc4_load_ebsd', { ebsdPath, material })
}

export async function tc4LoadSem(semImagePath: string): Promise<SemFeatures> {
  return invoke('tc4_load_sem', { semImagePath })
}

export async function tc4LoadLoad(loadPath: string): Promise<LoadFeatures> {
  return invoke('tc4_load_load', { loadPath })
}

export async function tc4MultimodalPredict(
  ebsd: EbsdFeatures,
  sem: SemFeatures,
  load: LoadFeatures
): Promise<MultimodalPrediction> {
  return invoke('tc4_multimodal_predict', { ebsd, sem, load })
}

export async function tc4MdValidation(config: Tc4MdValidationConfig): Promise<string> {
  return invoke('tc4_md_validation', { config })
}

// ============ Pinia Store ============

export const useTc4FailureStore = defineStore('tc4Failure', {
  state: () => ({
    // 数据状态
    ebsd: null as EbsdFeatures | null,
    sem: null as SemFeatures | null,
    load: null as LoadFeatures | null,
    prediction: null as MultimodalPrediction | null,

    // 状态
    isLoading: false,
    isGenerating: false,
    error: null as string | null,

    // 历史记录
    history: [] as MultimodalPrediction[],

    // MD 验证结果
    mdResult: null as string | null,
  }),

  getters: {
    hasAllInputs: (state) => state.ebsd !== null && state.sem !== null && state.load !== null,

    failureModeLabel: (state) => {
      if (!state.prediction) return ''
      const labels: Record<FailureMode, string> = {
        HCF: '高周疲劳',
        LCF: '低周疲劳',
        TMF: '热机械疲劳',
        CREEP: '蠕变',
        OVERLOAD: '过载',
        FOD: '外来物损伤',
      }
      return labels[state.prediction.failure_mode] || state.prediction.failure_mode
    },

    cycleRange: (state) => {
      if (!state.prediction) return { min: 0, max: 0 }
      const cycles = state.prediction.cycles
      if (cycles < 1000) return { min: 0, max: 10000 }
      if (cycles < 1e6) return { min: 0, max: 1e7 }
      return { min: 0, max: 1e10 }
    },
  },

  actions: {
    async generateSynthetic(config: Tc4SyntheticConfig) {
      this.isGenerating = true
      this.error = null
      try {
        const result = await tc4GenerateSynthetic(config)
        return result
      } catch (e) {
        this.error = String(e)
        throw e
      } finally {
        this.isGenerating = false
      }
    },

    async loadEbsd(ebsdPath: string, material: string = 'TC4') {
      this.isLoading = true
      this.error = null
      try {
        this.ebsd = await tc4LoadEbsd(ebsdPath, material)
        return this.ebsd
      } catch (e) {
        this.error = String(e)
        throw e
      } finally {
        this.isLoading = false
      }
    },

    async loadSem(semImagePath: string) {
      this.isLoading = true
      this.error = null
      try {
        this.sem = await tc4LoadSem(semImagePath)
        return this.sem
      } catch (e) {
        this.error = String(e)
        throw e
      } finally {
        this.isLoading = false
      }
    },

    async loadLoad(loadPath: string) {
      this.isLoading = true
      this.error = null
      try {
        this.load = await tc4LoadLoad(loadPath)
        return this.load
      } catch (e) {
        this.error = String(e)
        throw e
      } finally {
        this.isLoading = false
      }
    },

    async predict() {
      if (!this.ebsd || !this.sem || !this.load) {
        throw new Error('请先加载所有三路输入数据')
      }
      this.isLoading = true
      this.error = null
      try {
        this.prediction = await tc4MultimodalPredict(this.ebsd, this.sem, this.load)
        if (this.prediction) {
          this.history.unshift(this.prediction)
          if (this.history.length > 50) this.history.pop()
        }
        return this.prediction
      } catch (e) {
        this.error = String(e)
        throw e
      } finally {
        this.isLoading = false
      }
    },

    async runMdValidation(config: Tc4MdValidationConfig) {
      this.isLoading = true
      this.error = null
      try {
        this.mdResult = await tc4MdValidation(config)
        return this.mdResult
      } catch (e) {
        this.error = String(e)
        throw e
      } finally {
        this.isLoading = false
      }
    },

    clearInputs() {
      this.ebsd = null
      this.sem = null
      this.load = null
      this.prediction = null
      this.error = null
    },

    clearHistory() {
      this.history = []
    },
  },
})

File: src-tauri/src/commands/tc4_failure.rs
/**
 * tc4_failure.rs — V3.8 TC4 失效分析模块
 *
 * TC4 (Ti-6Al-4V) 钛合金多模态失效分析
 * 三路输入：EBSD 晶格 + SEM 图像 + Load 载荷
 * 输出：失效模式分类 + 疲劳寿命预测
 *
 * 复用 V3.6 SEM 分析和 V3.7 SurrogateModel
 */

use serde::{Deserialize, Serialize};

// ============ 类型定义 ============

/// EBSD 晶格特征 (80维)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EbsdFeatures {
    pub sample_id: String,
    pub features: Vec<f64>,
    // TC4 HCP 晶格参数
    pub a: Option<f64>,  // Å
    pub c: Option<f64>,  // Å
    pub c_to_a: Option<f64>,
}

/// SEM 图像特征 (768维 via ViT)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SemFeatures {
    pub sample_id: String,
    pub features: Vec<f64>,
    pub quality_score: f64,
}

/// Load 载荷特征 (30维)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LoadFeatures {
    pub sample_id: String,
    pub features: Vec<f64>,
    pub load_type: String,  // uniaxial / multiaxial / thermal
}

/// 失效模式
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum FailureMode {
    HCF,      // High Cycle Fatigue
    LCF,      // Low Cycle Fatigue
    TMF,      // Thermo-Mechanical Fatigue
    CREEP,    // Creep
    OVERLOAD, // Overload
    FOD,      // Foreign Object Damage
}

impl std::fmt::Display for FailureMode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            FailureMode::HCF => write!(f, "HCF"),
            FailureMode::LCF => write!(f, "LCF"),
            FailureMode::TMF => write!(f, "TMF"),
            FailureMode::CREEP => write!(f, "CREEP"),
            FailureMode::OVERLOAD => write!(f, "OVERLOAD"),
            FailureMode::FOD => write!(f, "FOD"),
        }
    }
}

/// 多模态预测结果
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MultimodalPrediction {
    pub sample_id: String,
    // 分类
    pub failure_mode: FailureMode,
    pub class_confidence: f64,  // 分类置信度
    // 回归
    pub log_cycles: f64,      // log10(cycles)
    pub cycles: f64,           // 预测寿命（秒）
    pub cycle_confidence: f64, // 回归置信度
    // 特征向量（用于后续分析）
    pub fused_features: Vec<f64>,
}

/// TC4 合成数据生成参数
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Tc4SyntheticConfig {
    pub num_samples: usize,
    pub output_dir: String,
    pub use_tc4: bool,  // true=TC4 HCP, false=Al FCC 先跑通
    pub seed: Option<u64>,
}

/// TC4 MD 验证参数
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Tc4MdValidationConfig {
    pub lattice_a: f64,  // Å
    pub lattice_c: f64,    // Å
    pub crack_plane: String,
    pub crack_direction: String,
    pub validation_type: String,  // crack_propagation / dislocation_evolution
}

// ============ 命令处理 ============

/**
 * tc4_generate_synthetic — 生成 TC4 合成数据
 *
 * 生成 4 个 CSV 文件：
 * - ebsd_synthetic.csv (晶格特征 80维)
 * - sem_synthetic.csv (图像特征 768维)
 * - load_synthetic.csv (载荷特征 30维)
 * - failure_labels.csv (失效标签)
 */
#[tauri::command]
pub async fn tc4_generate_synthetic(
    config: Tc4SyntheticConfig,
) -> Result<String, String> {
    println!("[TC4] 生成 {} 个合成样本", config.num_samples);

    // 调用 Python 合成数据生成器
    // 简化：这里直接生成简单的合成数据
    // 完整实现需要调用 Python 脚本

    Ok(format!(
        "TC4 synthetic data generation: {} samples -> {}",
        config.num_samples, config.output_dir
    ))
}

/**
 * tc4_load_ebsd — 加载 EBSD 晶格数据
 *
 * 从 .ctf 或 .ang 文件提取晶格特征 (80维)
 * 复用 orix 库
 */
#[tauri::command]
pub async fn tc4_load_ebsd(
    ebsd_path: String,
    material: String,
) -> Result<EbsdFeatures, String> {
    println!("[TC4] 加载 EBSD 数据: {} ({})", ebsd_path, material);

    // 简化：生成随机特征
    let features = (0..80).map(|_| rand_double()).collect();

    Ok(EbsdFeatures {
        sample_id: format!("EBSD_{}", uuid_simple()),
        features,
        a: Some(2.95),
        c: Some(4.68),
        c_to_a: Some(1.586),
    })
}

/**
 * tc4_load_sem — 加载 SEM 图像并提取特征
 *
 * 复用 V3.6 的 analyze_sem_image 工具
 */
#[tauri::command]
pub async fn tc4_load_sem(
    sem_image_path: String,
) -> Result<SemFeatures, String> {
    println!("[TC4] 加载 SEM 图像: {}", sem_image_path);

    // 简化：生成随机特征
    let features = (0..768).map(|_| rand_double()).collect();

    Ok(SemFeatures {
        sample_id: format!("SEM_{}", uuid_simple()),
        features,
        quality_score: 0.92,
    })
}

/**
 * tc4_load_load — 加载载荷历史数据
 *
 * 从 CSV 提取载荷特征 (30维)
 */
#[tauri::command]
pub async fn tc4_load_load(
    load_path: String,
) -> Result<LoadFeatures, String> {
    println!("[TC4] 加载载荷历史: {}", load_path);

    // 简化：生成随机特征
    let features = (0..30).map(|_| rand_double()).collect();

    Ok(LoadFeatures {
        sample_id: format!("LOAD_{}", uuid_simple()),
        features,
        load_type: "uniaxial".to_string(),
    })
}

/**
 * tc4_multimodal_predict — 多模态失效预测
 *
 * 三路输入：
 * - EBSD 晶格特征 (80维)
 * - SEM 图像特征 (768维)
 * - Load 载荷特征 (30维)
 *
 * 输出：
 * - 失效模式分类 (6类)
 * - 疲劳寿命预测 (log10(cycles))
 */
#[tauri::command]
pub async fn tc4_multimodal_predict(
    ebsd: EbsdFeatures,
    sem: SemFeatures,
    load: LoadFeatures,
) -> Result<MultimodalPrediction, String> {
    println!(
        "[TC4] 多模态预测: EBSD({}) + SEM({}) + LOAD({})",
        ebsd.sample_id, sem.sample_id, load.sample_id
    );

    // 简化：基于规则的预测
    // 真实场景需要调用训练好的 PyTorch 模型

    let temp = load.features.get(9).unwrap_or(&300.0);
    let freq = load.features.get(6).unwrap_or(&1.0);
    let stress_range = load.features.get(3).unwrap_or(&500.0);

    let (failure_mode, log_cycles, confidence) = if *temp > 500.0 {
        if rand_double() > 0.3 {
            ("TMF", 5.0 + rand_double(), 0.85)
        } else {
            ("CREEP", 4.0 + rand_double(), 0.80)
        }
    } else if *freq > 20.0 {
        ("HCF", 7.0 + rand_double() * 2.0, 0.88)
    } else if *stress_range > 500.0 {
        ("LCF", 4.0 + rand_double() * 2.0, 0.82)
    } else {
        ("LCF", 5.0 + rand_double() * 1.5, 0.75)
    };

    Ok(MultimodalPrediction {
        sample_id: format!("TC4_{}", uuid_simple()),
        failure_mode: parse_failure_mode(failure_mode),
        class_confidence: confidence,
        log_cycles,
        cycles: 10_f64.powf(log_cycles),
        cycle_confidence: confidence * 0.9,
        fused_features: vec![0.0; 256],  // 融合后 256维
    })
}

/**
 * tc4_md_validation — LAMMPS 分子动力学验证
 *
 * 复用 molecular_dynamics.rs 的 LAMMPS 接口
 * 模拟裂纹扩展和位错演化
 */
#[tauri::command]
pub async fn tc4_md_validation(
    config: Tc4MdValidationConfig,
) -> Result<String, String> {
    println!(
        "[TC4] MD 验证: {} / {} / {}",
        config.lattice_a, config.lattice_c, config.crack_plane
    );

    // 简化：返回模拟结果
    // 真实场景调用 LAMMPS

    Ok(format!(
        r#"{{
            "status": "completed",
            "crack_velocity": {:.4e},
            "dislocation_density": {:.4e},
            "fracture_energy": {:.4e},
            "trajectory_file": "{}/crack_trajectory.xyz"
        }}"#,
        rand_double() * 1e6,
        rand_double() * 1e15,
        rand_double() * 1e3,
        config.validation_type
    ))
}

// ============ 辅助函数 ============

fn rand_double() -> f64 {
    use std::time::{SystemTime, UNIX_EPOCH};
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .subsec_nanos();
    nanos as f64 / u32::MAX as f64
}

fn uuid_simple() -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .subsec_nanos();
    format!("{:08x}", nanos)
}

fn parse_failure_mode(s: &str) -> FailureMode {
    match s {
        "HCF" => FailureMode::HCF,
        "LCF" => FailureMode::LCF,
        "TMF" => FailureMode::TMF,
        "CREEP" => FailureMode::CREEP,
        "OVERLOAD" => FailureMode::OVERLOAD,
        "FOD" => FailureMode::FOD,
        _ => FailureMode::LCF,
    }
}